python/api/job/serializers.py

```diff
@@ -23,44 +21,61 @@
-        try:
-            if obj_type == 'cluster':
-                cluster = Cluster.objects.get(id=selector[obj_type])
-                name = cluster.name
-            elif obj_type == 'service':
-                service = ClusterObject.objects.get(id=selector[obj_type])
-                name = service.prototype.display_name
-            elif obj_type == 'component':
-                comp = ServiceComponent.objects.get(id=selector[obj_type])
-                name = comp.prototype.display_name
-            elif obj_type == 'provider':
-                provider = HostProvider.objects.get(id=selector[obj_type])
-                name = provider.name
-            elif obj_type == 'host':
-                host = Host.objects.get(id=selector[obj_type])
-                name = host.fqdn
-            else:
-                name = ''
-        except ObjectDoesNotExist:
-            name = 'does not exist'
-        resp.append(
-            {
-                'type': obj_type,
-                'id': selector[obj_type],
-                'name': name,
-            }
-        )
+def get_object_name(obj):
+    if hasattr(obj, 'name'):
+        return obj.name
+    elif hasattr(obj, 'fqdn'):
+        return obj.fqdn
+    else:
+        return obj.prototype.display_name
+
+
+def get_job_objects(task):
+    if not task.task_object:
+        return None
+
+    def cook_obj(obj_type, obj_id, name):
+        return {'type': obj_type, 'id': obj_id, 'name': name}
+
+    obj = task.task_object
+    obj_type = obj.prototype.type
+    resp = [cook_obj(obj_type, obj.id, get_object_name(obj))]
+    if obj_type == 'service':
+        resp.append(cook_obj('cluster', obj.cluster.id, get_object_name(obj.cluster)))
+    elif obj_type == 'component':
+        resp.append(cook_obj('cluster', obj.cluster.id, get_object_name(obj.cluster)))
+        resp.append(cook_obj('service', obj.service.id, get_object_name(obj.service)))
+    elif obj_type == 'host':
+        resp.append(cook_obj('provider', obj.provider.id, get_object_name(obj.provider)))
     return resp
 
 
+def get_task_selector(obj, action):
+    if not obj:
+        return None
+    selector = {obj.prototype.type: obj.id}
+    if obj.prototype.type == 'service':
+        selector['cluster'] = obj.cluster.id
+    if obj.prototype.type == 'component':
+        selector['cluster'] = obj.cluster.id
+        selector['service'] = obj.service.id
+    if isinstance(obj, Host) and action.host_action:
+        cluster = get_object_cluster(obj)
+        if action.prototype.type == 'component':
+            component = ServiceComponent.obj.get(cluster=cluster, prototype=action.prototype)
+            selector['component'] = component.id
+        if action.prototype.type == 'service':
+            service = ClusterObject.obj.get(cluster=cluster, prototype=action.prototype)
+            selector['service'] = service.id
+        if obj.cluster is not None:
+            selector['cluster'] = obj.cluster.id
```

Comment:
```suggestion
    if isinstance(obj, Host) and action.host_action:
        if action.prototype.type == 'component':
            service = ClusterObject.obj.get(prototype=action.prototype.parent, cluster=obj.cluster)
            component = ServiceComponent.obj.get(
                cluster=obj.cluster, service=service, prototype=action.prototype
            )
            selector['service'] = service.id
            selector['component'] = component.id
        if action.prototype.type == 'service':
            service = ClusterObject.obj.get(prototype=action.prototype, cluster=obj.cluster)
            selector['service'] = service.id
        if obj.cluster is not None:
            selector['cluster'] = obj.cluster.id
```
The block for the component type lacks a service in the selector, and there is also no receiving a component for a specific service.